File: apps/reading/src/api/fullscreenDraftCollectivePresentedTwinsCompose.ts
/**
 * Floating fullscreen-open over draft-before-merge collective presented twins
 * pack (pure).
 *
 * Operator vision: after provisional combined draft + multi-select collective
 * of presented twins, open a floating deep-research instance fullscreen —
 * without live dispatch or parent merge.
 *
 * live_dispatched / merge_executed / draft_written always false.
 * purchase_executed / live_router_authorized always false.
 * production_router_verdict always REJECT.
 */

import {
  composeFloatingFullscreenOpen,
  type FloatingFullscreenOpenCompose,
  type FloatingFullscreenOpenInput,
} from "./floatingFullscreenOpenCompose";
import {
  composeDraftBeforeMergeCollectivePresentedTwins,
  type DraftBeforeMergeCollectivePresentedTwinsCompose,
  type DraftBeforeMergeCollectivePresentedTwinsInput,
} from "./draftBeforeMergeCollectivePresentedTwinsCompose";

export interface FullscreenDraftCollectivePresentedTwinsInput {
  fullscreen: Omit<FloatingFullscreenOpenInput, "operator_ack">;
  draft_collective: Omit<
    DraftBeforeMergeCollectivePresentedTwinsInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface FullscreenDraftCollectivePresentedTwinsCompose {
  session_id: string;
  parent_asset_id: string;
  title: string;
  account_id: string;
  week_id: string;
  asset_id: string;
  fullscreen: FloatingFullscreenOpenCompose;
  draft_collective: DraftBeforeMergeCollectivePresentedTwinsCompose;
  pack_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  draft_written: false;
  analysis_written: false;
  purchase_executed: false;
  charge_executed: false;
  hosted: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  live_router_authorized: false;
  twin_written: false;
  prompts_injected: false;
  live_dispatch_authorized: false;
  live_execution_authorized: false;
  remote_fetched: false;
  backlog_mutated: false;
  secrets_stored: false;
  live_meter_read: false;
  store_mutated: false;
  suite_rewritten: false;
  remote_index_queried: false;
  inventory_mutated: false;
  record_persisted: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "fullscreen_draft_collective_presented_twins_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Fullscreen-open stacked on draft-before-merge collective pack.
 * Never live-dispatches; never merges; never purchases; ND REJECT.
 */
export function composeFullscreenDraftCollectivePresentedTwins(
  input: FullscreenDraftCollectivePresentedTwinsInput,
): FullscreenDraftCollectivePresentedTwinsCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.fullscreen || typeof input.fullscreen !== "object") {
    throw new Error("fullscreen must be an object");
  }
  if (!input.draft_collective || typeof input.draft_collective !== "object") {
    throw new Error("draft_collective must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · merge_executed=false · draft_written=false",
    "purchase_executed=false · live_router_authorized=false",
    "production_router_verdict=REJECT",
  ];

  const fullscreen = composeFloatingFullscreenOpen({
    ...input.fullscreen,
    operator_ack: input.operator_ack,
  });
  notes.push(...fullscreen.notes.map((n) => `[fullscreen] ${n}`));

  const draft_collective = composeDraftBeforeMergeCollectivePresentedTwins({
    ...input.draft_collective,
    operator_ack: input.operator_ack,
  });
  notes.push(
    ...draft_collective.notes.map((n) => `[draft_collective] ${n}`),
  );

  const session_id = requireNonEmpty(fullscreen.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    fullscreen.parent_asset_id,
    "parent_asset_id",
  );
  const title = requireNonEmpty(draft_collective.title, "title");
  const account_id = requireNonEmpty(
    draft_collective.account_id,
    "account_id",
  );
  const week_id = requireNonEmpty(draft_collective.week_id, "week_id");
  const asset_id = requireNonEmpty(draft_collective.asset_id, "asset_id");

  const session_aligned = draft_collective.session_id === session_id;
  const parent_aligned = draft_collective.parent_asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between fullscreen and draft_collective — pack_ready blocked",
    );
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between fullscreen and draft_collective — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      fullscreen.fullscreen_ready === true &&
      draft_collective.pack_ready === true &&
      fullscreen.live_dispatched === false &&
      fullscreen.merge_executed === false &&
      fullscreen.pack_dispatched === false &&
      draft_collective.draft_written === false &&
      draft_collective.merge_executed === false &&
      draft_collective.live_dispatched === false &&
      draft_collective.purchase_executed === false &&
      draft_collective.live_router_authorized === false &&
      draft_collective.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      fullscreen.merge_executed === false &&
      draft_collective.purchase_executed === false &&
      draft_collective.production_router_verdict === "REJECT" &&
      (fullscreen.fullscreen_ready === true ||
        draft_collective.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — fullscreen + draft-before-merge collective pack ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — fullscreen, draft_collective, alignment, or operator_ack gate open",
    );
  }

  if (
    fullscreen.live_dispatched !== false ||
    fullscreen.merge_executed !== false ||
    fullscreen.pack_dispatched !== false ||
    draft_collective.draft_written !== false ||
    draft_collective.merge_executed !== false ||
    draft_collective.live_dispatched !== false ||
    draft_collective.purchase_executed !== false ||
    draft_collective.live_router_authorized !== false ||
    draft_collective.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("purchase_executed=false");
  notes.push("charge_executed=false");
  notes.push("hosted=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("live_router_authorized=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("live_execution_authorized=false");
  notes.push("remote_fetched=false");
  notes.push("backlog_mutated=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("remote_index_queried=false");
  notes.push("inventory_mutated=false");
  notes.push("record_persisted=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    title,
    account_id,
    week_id,
    asset_id,
    fullscreen,
    draft_collective,
    pack_ready,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    draft_written: false,
    analysis_written: false,
    purchase_executed: false,
    charge_executed: false,
    hosted: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    live_router_authorized: false,
    twin_written: false,
    prompts_injected: false,
    live_dispatch_authorized: false,
    live_execution_authorized: false,
    remote_fetched: false,
    backlog_mutated: false,
    secrets_stored: false,
    live_meter_read: false,
    store_mutated: false,
    suite_rewritten: false,
    remote_index_queried: false,
    inventory_mutated: false,
    record_persisted: false,
    production_router_verdict: "REJECT",
    notes,
    authority: "fullscreen_draft_collective_presented_twins_compose_advisory",
  };
}

export function formatFullscreenDraftCollectivePresentedTwinsSummary(
  c: FullscreenDraftCollectivePresentedTwinsCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `fullscreen_ready=${c.fullscreen.fullscreen_ready} · ` +
    `draft_collective_ready=${c.draft_collective.pack_ready} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_dispatched=false · merge_executed=false · draft_written=false`
  );
}

File: substrate/fullscreen_draft_collective_presented_twins_compose.py
"""Fullscreen-open over draft-before-merge collective presented twins pack.

live_dispatched / merge_executed / draft_written always False.
purchase_executed / live_router_authorized always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.draft_before_merge_collective_presented_twins_compose import (
    DraftBeforeMergeCollectivePresentedTwinsCompose,
    DraftBeforeMergeCollectivePresentedTwinsComposeError,
    compose_draft_before_merge_collective_presented_twins,
)
from substrate.floating_fullscreen_open_compose import (
    FloatingFullscreenOpenCompose,
    FloatingFullscreenOpenComposeError,
    compose_floating_fullscreen_open,
)


class FullscreenDraftCollectivePresentedTwinsComposeError(ValueError):
    """Fail-closed validation for fullscreen + draft collective pack."""


@dataclass(frozen=True)
class FullscreenDraftCollectivePresentedTwinsCompose:
    session_id: str
    parent_asset_id: str
    title: str
    account_id: str
    week_id: str
    asset_id: str
    fullscreen: FloatingFullscreenOpenCompose
    draft_collective: DraftBeforeMergeCollectivePresentedTwinsCompose
    pack_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    draft_written: bool
    analysis_written: bool
    purchase_executed: bool
    charge_executed: bool
    hosted: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    live_router_authorized: bool
    twin_written: bool
    prompts_injected: bool
    live_dispatch_authorized: bool
    live_execution_authorized: bool
    remote_fetched: bool
    backlog_mutated: bool
    secrets_stored: bool
    live_meter_read: bool
    store_mutated: bool
    suite_rewritten: bool
    remote_index_queried: bool
    inventory_mutated: bool
    record_persisted: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "fullscreen": self.fullscreen.to_dict(),
            "draft_collective": self.draft_collective.to_dict(),
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "draft_written": False,
            "analysis_written": False,
            "purchase_executed": False,
            "charge_executed": False,
            "hosted": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "live_router_authorized": False,
            "twin_written": False,
            "prompts_injected": False,
            "live_dispatch_authorized": False,
            "live_execution_authorized": False,
            "remote_fetched": False,
            "backlog_mutated": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "remote_index_queried": False,
            "inventory_mutated": False,
            "record_persisted": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "fullscreen_draft_collective_presented_twins_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_fullscreen_draft_collective_presented_twins(
    *,
    fullscreen: object,
    draft_collective: object,
    operator_ack: object,
    require_both: object | None = None,
) -> FullscreenDraftCollectivePresentedTwinsCompose:
    """Fullscreen-open + draft collective pack. Never dispatches/merges."""
    if not isinstance(operator_ack, bool):
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(fullscreen, dict):
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            "fullscreen must be an object"
        )
    if not isinstance(draft_collective, dict):
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            "draft_collective must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · merge_executed=false · draft_written=false",
        "purchase_executed=false · live_router_authorized=false",
        "production_router_verdict=REJECT",
    ]

    try:
        fs = compose_floating_fullscreen_open(
            session_id=fullscreen.get("session_id"),
            parent_asset_id=fullscreen.get("parent_asset_id"),
            operator_ack=operator_ack,
            existing_instance=fullscreen.get("existing_instance"),
            highlight=fullscreen.get("highlight"),
            prompt=fullscreen.get("prompt"),
            gated=fullscreen.get("gated"),
            tray_siblings=fullscreen.get("tray_siblings"),
        )
    except FloatingFullscreenOpenComposeError as e:
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            str(e)
        ) from e
    notes.extend(f"[fullscreen] {n}" for n in fs.notes)

    try:
        dc = compose_draft_before_merge_collective_presented_twins(
            draft_gate=draft_collective.get("draft_gate"),
            collective_pack=draft_collective.get("collective_pack"),
            operator_ack=operator_ack,
            require_both=draft_collective.get("require_both"),
        )
    except DraftBeforeMergeCollectivePresentedTwinsComposeError as e:
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            str(e)
        ) from e
    notes.extend(f"[draft_collective] {n}" for n in dc.notes)

    session = _require_nonempty(fs.session_id, field="session_id")
    parent = _require_nonempty(fs.parent_asset_id, field="parent_asset_id")
    title = _require_nonempty(dc.title, field="title")
    account = _require_nonempty(dc.account_id, field="account_id")
    week = _require_nonempty(dc.week_id, field="week_id")
    asset = _require_nonempty(dc.asset_id, field="asset_id")

    session_aligned = dc.session_id == session
    parent_aligned = dc.parent_asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between fullscreen and draft_collective — "
            "pack_ready blocked"
        )
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between fullscreen and draft_collective — "
            "pack_ready blocked"
        )

    if require:
        pack_ready = (
            session_aligned
            and parent_aligned
            and fs.fullscreen_ready is True
            and dc.pack_ready is True
            and fs.live_dispatched is False
            and fs.merge_executed is False
            and fs.pack_dispatched is False
            and dc.draft_written is False
            and dc.merge_executed is False
            and dc.live_dispatched is False
            and dc.purchase_executed is False
            and dc.live_router_authorized is False
            and dc.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned
            and parent_aligned
            and operator_ack is True
            and fs.merge_executed is False
            and dc.purchase_executed is False
            and dc.production_router_verdict == "REJECT"
            and (fs.fullscreen_ready is True or dc.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — fullscreen + draft-before-merge collective pack "
            "ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — fullscreen, draft_collective, alignment, or "
            "operator_ack gate open"
        )

    if (
        fs.live_dispatched is not False
        or fs.merge_executed is not False
        or fs.pack_dispatched is not False
        or dc.draft_written is not False
        or dc.merge_executed is not False
        or dc.live_dispatched is not False
        or dc.purchase_executed is not False
        or dc.live_router_authorized is not False
        or dc.production_router_verdict != "REJECT"
    ):
        raise FullscreenDraftCollectivePresentedTwinsComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "draft_written=false",
            "analysis_written=false",
            "purchase_executed=false",
            "charge_executed=false",
            "hosted=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "live_router_authorized=false",
            "twin_written=false",
            "prompts_injected=false",
            "live_dispatch_authorized=false",
            "live_execution_authorized=false",
            "remote_fetched=false",
            "backlog_mutated=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "remote_index_queried=false",
            "inventory_mutated=false",
            "record_persisted=false",
            "production_router_verdict=REJECT",
        )
    )

    return FullscreenDraftCollectivePresentedTwinsCompose(
        session_id=session,
        parent_asset_id=parent,
        title=title,
        account_id=account,
        week_id=week,
        asset_id=asset,
        fullscreen=fs,
        draft_collective=dc,
        pack_ready=pack_ready,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        draft_written=False,
        analysis_written=False,
        purchase_executed=False,
        charge_executed=False,
        hosted=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        live_router_authorized=False,
        twin_written=False,
        prompts_injected=False,
        live_dispatch_authorized=False,
        live_execution_authorized=False,
        remote_fetched=False,
        backlog_mutated=False,
        secrets_stored=False,
        live_meter_read=False,
        store_mutated=False,
        suite_rewritten=False,
        remote_index_queried=False,
        inventory_mutated=False,
        record_persisted=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "fullscreen_draft_collective_presented_twins_compose_advisory"
        ),
    )


def format_fullscreen_draft_collective_presented_twins_summary(
    c: FullscreenDraftCollectivePresentedTwinsCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"fullscreen_ready={c.fullscreen.fullscreen_ready} · "
        f"draft_collective_ready={c.draft_collective.pack_ready} · "
        f"verdict={c.production_router_verdict} · "
        f"live_dispatched=false · merge_executed=false · draft_written=false"
    )


__all__ = [
    "FullscreenDraftCollectivePresentedTwinsCompose",
    "FullscreenDraftCollectivePresentedTwinsComposeError",
    "compose_fullscreen_draft_collective_presented_twins",
    "format_fullscreen_draft_collective_presented_twins_summary",
]
